scheduler/src/cook/kubernetes/api.clj

```diff
@@ -0,0 +1,316 @@
+          job-status (first (filter (fn [c] (= cook-container-name-for-job (.getName c)))
+                                    container-statuses))]
+      (if (some-> pod .getMetadata .getDeletionTimestamp)
+        ; If a pod has been ordered deleted, treat it as if it was gone, Its being async removed.
+        {:state :missing :reason "Pod was explicitly deleted"}
+        ; If pod isn't being async removed, then look at the containers inside it....
+        (if job-status
+          (let [^V1ContainerState state (.getState job-status)]
+            (cond
+              (.getWaiting state)
+              {:state :pod/waiting
+               :reason (-> state .getWaiting .getReason)}
+              (.getRunning state)
+              {:state :pod/running
+               :reason "Running"}
+              (.getTerminated state)
+              (let [exit-code (-> state .getTerminated .getExitCode)]
+                (if (= 0 exit-code)
+                  {:state :pod/succeeded
+                   :exit exit-code
+                   :reason (-> state .getTerminated .getReason)}
+                  {:state :pod/failed
+                   :exit exit-code
+                   :reason (-> state .getTerminated .getReason)}))
+              :default
+              {:state :pod/unknown
+               :reason "Unknown"}))
+
+          {:state :pod/waiting
+           :reason "Pending"})))))
+
+(defn kill-task
+  "Kill this kubernetes pod"
+  [^ApiClient api-client ^V1Pod pod]
+  (let [api (CoreV1Api. api-client)
+        ^V1DeleteOptions deleteOptions (-> (V1DeleteOptionsBuilder.) (.withPropagationPolicy "Background") .build)]
+    (try
+      (.deleteNamespacedPod
+        api
+        (-> pod .getMetadata .getName)
+        (-> pod .getMetadata .getNamespace)
+        deleteOptions
+        nil
+        nil
+        nil
+        nil
+        nil)
+      (catch JsonSyntaxException e
+        (log/info "Caught the https://github.com/kubernetes-client/java/issues/252 exception.")
+        ; Silently gobble this exception.
+        ;
+        ; The java API can throw a a JsonSyntaxException parsing the kubernetes reply!
+        ; https://github.com/kubernetes-client/java/issues/252
+        ; https://github.com/kubernetes-client/java/issues/86
+        ;
+        ; They actually advise catching the exception and moving on!
+        ;
+        ))))
+
+
+(defn remove-finalization-if-set-and-delete
+  "Remove finalization for a pod if its there. No-op if its not there.
+  We always delete unconditionally, so finalization doesn't matter."
+  [^ApiClient api-client ^V1Pod pod]
+  ; TODO: This likes to noisily throw NotFound multiple times as we delete away from kubernetes.
+  ; I suspect our predicate of existing-states-equivalent needs tweaking.
+  (kill-task api-client pod))
+
+(defn launch-task
+  "Given a pod-name use lookup the associated task, extract the parts needed to syntehsize the kubenretes object and go"
```

Comment:
Grammar doesn't make sense in this docstring

Reply:
Changed.